### PR TITLE
[risk=no][RW-8975] Added pulse check for API proxy.

### PR DESCRIPTION
**Pull request by evrii**
Description:
Added pulse check for API proxy to new e2e2 tests.
In order to test this, either do not start the api proxy when running the e2ev2 tests or start it on a port other than 8080.

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: ui/src/app/pages/analysis/new-notebook-modal.tsx [hunk 1]
```diff
@@ -33,7 +34,7 @@ export const NewNotebookModal = (props: Props) => {
   const [nameTouched, setNameTouched] = useState(false);
   const [navigate] = useNavigation();
 
-  const { workspace, onClose, existingNameList } = props;
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/new-notebook-modal.tsx:36]: These files should not be here - something has gone wrong.  Could you fetch and rebase against the latest `main` ?  I can pair on this.

**Review by jmthibault79 — COMMENTED** [review]
This does not seem to work for me.  What I see:
```
thibault@wm98d-e39 ~/src/aou/e2ev2 (eric/RW-8975) $ yarn test tests/sanity.browser.test.js
yarn run v1.22.19
$ jest tests/sanity.browser.test.js
 FAIL  tests/sanity.browser.test.js (5.25 s)
  ✓ page loads (2363 ms)
  ✕ view cookie policy page (672 ms)
  ✕ navigation to /workspaces redirects to sign-in page (671 ms)
  ✕ navigation to /profile redirects to sign-in page (661 ms)

  ● view cookie policy page

    connect ECONNREFUSED 127.0.0.1:8080



  ● navigation to /workspaces redirects to sign-in page

    connect ECONNREFUSED 127.0.0.1:8080



  ● navigation to /profile redirects to sign-in page

    connect ECONNREFUSED 127.0.0.1:8080



Test Suites: 1 failed, 1 total
Tests:       3 failed, 1 passed, 4 total
Snapshots:   0 total
Time:        5.319 s
Ran all test suites matching /tests\/sanity.browser.test.js/i.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: ui/src/app/pages/analysis/new-notebook-modal.tsx (hunk 1 above)
Comment (evrii) [reply ui/src/app/pages/analysis/new-notebook-modal.tsx]: Good call. I merged main into the branch, so I believe that this should be fixed now.

**Comment by evrii** [comment]
> This does not seem to work for me. What I see:
> 
> ```
> thibault@wm98d-e39 ~/src/aou/e2ev2 (eric/RW-8975) $ yarn test tests/sanity.browser.test.js
> yarn run v1.22.19
> $ jest tests/sanity.browser.test.js
>  FAIL  tests/sanity.browser.test.js (5.25 s)
>   ✓ page loads (2363 ms)
>   ✕ view cookie policy page (672 ms)
>   ✕ navigation to /workspaces redirects to sign-in page (671 ms)
>   ✕ navigation to /profile redirects to sign-in page (661 ms)
> 
>   ● view cookie policy page
> 
>     connect ECONNREFUSED 127.0.0.1:8080
> 
> 
> 
>   ● navigation to /workspaces redirects to sign-in page
> 
>     connect ECONNREFUSED 127.0.0.1:8080
> 
> 
> 
>   ● navigation to /profile redirects to sign-in page
> 
>     connect ECONNREFUSED 127.0.0.1:8080
> 
> 
> 
> Test Suites: 1 failed, 1 total
> Tests:       3 failed, 1 passed, 4 total
> Snapshots:   0 total
> Time:        5.319 s
> Ran all test suites matching /tests\/sanity.browser.test.js/i.
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> ```

This looks to me like your api proxy is not running.

Does it work if you run the following in a different terminal from the e2ev2 directory:
```
cd ../api-proxy
PROXY_PORT=8080 PROXY_MODE=replay-only node startproxy.mjs
```

**Comment by jmthibault79** [comment]
> This looks to me like your api proxy is not running.

I thought the point of this change was to detect when the API proxy was not running ?  Seems like I am misunderstanding something :)

File: e2ev2/src/test-utils.js [hunk 2]
```diff
@@ -73,13 +74,25 @@ export_({browserTest})
...
+    })
+  }).then(async () => {
+    await page.setJavaScriptEnabled(false)
+    await page.goto(config.urlRoot(), {waitUntil: 'networkidle0'})
+    await page.evaluate(
+      url => { localStorage.allOfUsApiUrlOverride = url},
+      process.env.API_PROXY_URL
+    )
+    await page.setJavaScriptEnabled(true)
+  });
```
Comment (dmohs) [inline e2ev2/src/test-utils.js:94]:
```suggestion
  })
  await page.setJavaScriptEnabled(false)
  await page.goto(config.urlRoot(), {waitUntil: 'networkidle0'})
  await page.evaluate(
    url => { localStorage.allOfUsApiUrlOverride = url},
    process.env.API_PROXY_URL
  )
  await page.setJavaScriptEnabled(true)
```

File: e2ev2/tests/sanity.browser.test.js [hunk 3]
```diff
@@ -13,19 +13,24 @@ browserTest('page loads', async browser => {
...
+      expect(cpLink).toBeDefined()
+      // click and wait need to happen simultaneously to avoid a race
+      const [, target] = await Promise.all([
+        cpLink.click(),
+        browser.waitForTarget(target => target.opener() === page.target(), {timeout: 2e3})
+      ])
+      const cpPage = await target.page()
+      cpPage.setDefaultTimeout(2000)
+      const h3 = await cpPage.waitForSelector('h3')
+      expect(await h3.evaluate(n => n.innerText))
```
Comment (dmohs) [inline e2ev2/tests/sanity.browser.test.js:28]: Look for usage of expect(...).resolves.

**Comment by evrii** [comment]
> > This looks to me like your api proxy is not running.
> 
> I thought the point of this change was to detect when the API proxy was not running ? Seems like I am misunderstanding something :)

Whoops, you are absolutely right. I must have been looking at this for too long. The ECONNREFUSED error indicates the API Proxy is down, but that error is not telling you anything about the root cause. I updated the code with a more descriptive error message.

**Review by jmthibault79 — COMMENTED** [review]
Tested on M1 and Intel-Mac.  In both cases:

good: I now see "is your proxy running" when it is not running
bad: I still see this when I run the proxy.

My best guess is that the proxy port setting is no longer respected.  why would this be?
```
$ PROXY_PORT=8080 PROXY_MODE=replay-only node startproxy.mjs
logging to: stderr
server listening on port 52917.
```

**Comment by evrii** [comment]
> Tested on M1 and Intel-Mac. In both cases:
> 
> good: I now see "is your proxy running" when it is not running bad: I still see this when I run the proxy.
> 
> My best guess is that the proxy port setting is no longer respected. why would this be?
> 
> ```
> $ PROXY_PORT=8080 PROXY_MODE=replay-only node startproxy.mjs
> logging to: stderr
> server listening on port 52917.
> ```

Sorry, I had forgotten to update the README, but the api-proxy had changed a little while back. Now instead of an environment, startproxy.mjs expects the port as an argument.

File: e2ev2/tests/sanity.browser.test.js [hunk 4]
```diff
@@ -13,19 +13,24 @@ browserTest('page loads', async browser => {
...
-    cpLink.click(),
-    browser.waitForTarget(target => target.opener() === page.target(), {timeout: 2e3})
-  ])
-  const cpPage = await target.page()
-  cpPage.setDefaultTimeout(2000)
-  const h3 = await cpPage.waitForSelector('h3')
-  expect(await h3.evaluate(n => n.innerText))
-    .toBe('All of Us Research Program Cookie Policy')
+    .then(async () => await page.goto(config.urlRoot()+'/login', {waitUntil: 'networkidle0'}))
+    .then(async () => {
```
Comment (jmthibault79) [inline e2ev2/tests/sanity.browser.test.js:17]: What is the logic for splitting up the `await` calls in 2 `then`s like this?  Naively I would either expect a single `then` or many `then`s

**Review by jmthibault79 — APPROVED** [review]
It works now!  Thanks for this improvement.

File: e2ev2/tests/sanity.browser.test.js (hunk 4 above)
Comment (evrii) [reply e2ev2/tests/sanity.browser.test.js:16]: I believe that you are right, one then seems like the more intuitive approach. I updated it.

**Pull request merged**